src/memo/memo.ml

```diff
@@ -647,6 +666,55 @@ module Call_stack = struct
     Fiber.Var.set call_stack_var stack (fun () -> Implicit_output.forbid f)
 end
 
+module Error_handler : sig
+  val is_set : bool Fiber.t
+
+  val report_error : Exn_with_backtrace.t -> unit Fiber.t
+
+  val with_error_handler :
+    (Exn_with_backtrace.t -> unit Fiber.t) -> (unit -> 'a Fiber.t) -> 'a Fiber.t
+end = struct
+  type t = Exn_with_backtrace.t -> unit Fiber.t
+
+  let var : t Fiber.Var.t = Fiber.Var.create ()
+
+  let is_set = Fiber.map (Fiber.Var.get var) ~f:Option.is_some
+
+  let get_exn = Fiber.Var.get_exn var
+
+  let report_error error =
+    let open Fiber.O in
+    let* handler = get_exn in
+    let* stack = Call_stack.get_call_stack_without_state () in
+    let error =
+      Exn_with_backtrace.map error ~f:(fun exn ->
+          List.fold_left stack ~init:exn ~f:(fun exn stack_frame ->
+              Error.extend_stack exn ~stack_frame))
+    in
+    handler error
+
+  let deduplicate_errors f =
+    let reported = ref Exn_set.empty in
```

Comment:
It seems worth adding a comment here explaining how this interacts with the incremental mode: presumably, we're going to allocate a fresh empty set on every rebuild but it's not immediately obvious that this does happen from the surrounding code.

Reply:
Added a comment to with_error_handler:

    (* [with_error_handler] runs once for every incremental run, so calling
       [deduplicate_errors] afresh here makes sure that we re-report all errors*)